ActiveRecord::Base#find(array) returning result in the same order as the array passed

miguelgraz opened the pull request:
This is an attempt to change the default order of `ActiveRecord::Base#find(array)` according to [this issue](https://github.com/rails/rails/issues/20338).

Here's a list of concerns of the current implementation and a couple of doubts on how good it is right now and if there are other things to handle that aren't covered yet.
- [x] it should return the same order as the parameter, regardless if it a list or array of integer or strings or slugs
- [x] it should obey the `limit` clause
- [x] it should allow any `order` clause take precedence, making it backward compatible
- [ ] should it worry about (default|name)scope? How to test that?
- [ ] should it worry about the `offset` clause? (currently failing on that)
- [x] is `ActiveRecord::Relation.values[:order]` a good way to know if this relation was already ordered? Fixed using `order_values`, thanks to @bogdan 

I'm really not sure if that's the best way to achieve this feature behavior and would love all feedbacks!

Thank you! :snowboarder:

miguelgraz committed:
[#20338] WIP: first basic implementation and specs

miguelgraz committed:
[#20338] improving implementation, fixing and adding some more specs

miguelgraz committed:
[#20338] adding tests to ensure the order clause takes precedence

miguelgraz committed:
adding a test to ensure the find is obbeying the limit

miguelgraz committed:
do not change the order of the result if the object was already ordered by the user via :order clause

miguelgraz commented:
Perhaps I should have squashed these commits, will do it if/when the implementation is okay to be merged :smiley:

jmbejar commented on activerecord/lib/active_record/relation/finder_methods.rb:
```diff
@@ -459,7 +459,9 @@ def find_some(ids)
-        result
+        return result if self.values[:order]
+        records_by_id = result.index_by(&:id)
+        ids.first(expected_size).collect { |id| records_by_id[id.to_i] }
```
@miguelgraz How would it work when `limit_value` is present?

For example, if we pass a list of 5 ids but the limit is 2, I understand the query will retrieve only 2 records. Since we're passing the complete list of ids to the `where` condition, we don't know in advance which ids are associated with these 2 records. 

In other words, I think we can not deduce that `ids.first(expected_size)` are the corresponding ones when `expected_size` is less than `ids.size` (when `limit` is used in the relation).

miguelgraz replied:
Good call @jmbejar, it makes sense to me that it should work fine using `.limit` too, I'll work further on that.

Thanks for the heads-up and the test! :)

jmbejar commented on activerecord/test/cases/finder_test.rb:
```diff
@@ -48,6 +48,48 @@ def test_find_with_proc_parameter_and_block
+    assert_equal 'The Fifth Topic of the day', records[2].title
+  end
+
+  def test_find_with_ids_and_limit
```
Looks like I was able to confirm the issue commented in my previous comment changing a little bit this test

```ruby
  def test_find_with_ids_and_limit
    records = Topic.limit(1).find([4,2,5])
    assert_equal 1, records.size
    assert_equal 'The Fourth Topic of the day', records[0].title # Failure here, records is equals to `[nil]`
  end
```

Running this test in my machine, it seems that the topic with id=5 is retrieved but the new implementation looks for ids 4 and 2, so the result array has only a nil value. Of course, this is kind of random, so it could be hard to reproduce.

miguelgraz committed:
WIP: fixing the limit bug and introducing new tests (failing for now) on .find(array) with offset

miguelgraz commented on activerecord/lib/active_record/relation/finder_methods.rb:
```diff
@@ -456,10 +454,16 @@ def find_some(ids)
       if offset_value && (ids.size - offset_value < expected_size)
         expected_size = ids.size - offset_value
+      else
+        ids = ids.first(expected_size) unless self.values[:order]
```
@jmbejar you were completely right about the `.limit` problem, basically the `where` was returning records ordered by ID and we couldn't assume all of needed records will be there, so now this line aims to send to `where` all and only the needed IDs.
I also improved the `test_find_with_ids_and_limit` test using your idea and it seems to work better now.

Would love to have your review and opinion, thanks!

bogdan replied:
I think `order_values.empty?` is better here. I saw it in other places in code and it is the fastest method.

miguelgraz replied:
Awesome, I didn't know about it and wasn't happy using `self.values` hash :+1:

miguelgraz commented:
So even though I still want to improve this code a bit (hate to have two conditions on `self.values[:order]`) it seems to be working fine EXCEPT for requests with the `offset` clause. 
I just improved a bit the `test_find_by_ids_with_limit_and_offset` test and also added another one to ensure that `Topic.offset(2).find([3,2,5,1,4])` will return the same thing as `Topic.find([5,1,4])`. Is this behavior correct?

miguelgraz committed:
using order_values method instead of relying on ActiveRecord::Relation @values hash

jmbejar commented:
@miguelgraz Thinking again about this issue here.

I think we should simplify the approach for the case where `find(array)` and `limit` or `offset` are used at the same time. I'd simply pass the entire list of ids into the query and leave the database to determine the records to be returned.

Of course, the database will select the records using offset and limit and we can not guarantee that the first ids of the array are being to be related with the returned records. However, this is how DB works if we create a query with such condition and limit/ofsset values.

You've already added the line that should fix the whole thing:

`ids.collect { |id| records_by_id[id.to_i] }.compact`

where ids is exactly the same array passed to the `find` method. In this way, we are making sure that returned records are ordered in memory based on the provided ids array. I feel like we should not do extra calculations to cover the `limit`/`offset` case, it should be good enough.

I tested this implementation and it seems to work:

```ruby
    def find_some(ids)
      result = where(primary_key => ids).to_a

      expected_size =
        if limit_value && ids.size > limit_value
          limit_value
        else
          ids.size
        end

      # 11 ids with limit 3, offset 9 should give 2 results.
      if offset_value && (ids.size - offset_value < expected_size)
        expected_size = ids.size - offset_value
      end

      if result.size == expected_size
        return result if order_values.present?
        records_by_id = result.index_by(&:id)
        ids.collect { |id| records_by_id[id.to_i] }.compact
      else
        raise_record_not_found_exception!(ids, result.size, expected_size)
      end
    end
```


Regarding testing, assuming you take my suggested implementation for the `finf_some` method, the implementation of `test_find_with_ids_and_limit` and `test_find_with_ids_and_offset` is kind of tricky, because records being returned by the database are not deterministic in those cases.

For example, the query `select id from entrants where id in (1,2,3,4,5) limit 3` will return 3 records, but we don't now which ones. Results with ids (2,3,5) or (3,4,5) are both valid results, while a result with ids (3,4,1) is invalid because is not respecting the order of the original array of ids.

The following is a way to make those tests deterministic, but I know it's complicated code. Maybe someone comes up with a better (and simpler) idea?

```ruby
  def test_find_with_ids_and_limit
    records = Topic.limit(3).find([3,2,5,1,4])
    assert_equal 3, records.size
    assert_equal [3,2,5,1,4] && records.map(&:id), records.map(&:id)
  end

  def test_find_with_ids_and_offset
    records = Topic.offset(2).find([3,2,5,1,4])
    assert_equal 3, records.size
    assert_equal [3,2,5,1,4] && records.map(&:id), records.map(&:id)
  end
```

I tested all this code and tests are green :). Hope this suggestions help you to move forward this PR!

matthewd commented:
> I'd simply pass the entire list of ids into the query and leave the database to determine the records to be returned.

:+1:

> The following is a way to make those tests deterministic, but I know it's complicated code. Maybe someone comes up with a better (and simpler) idea?

Add an `order`?

miguelgraz commented:
Sorry for my late response, getting back to this. 

Thanks @jmbejar and @matthewd for your really valuable suggestions, I guess @jmbejar approach is fine and I'd be happy with it but since this PR came from this issue https://github.com/rails/rails/issues/20338 I'd like to know @dhh's opinion on this whole `order` with `offset` stuff as well.

I guess it'd be good to discuss this further and agree to pick an approach decided by "everyone" (or at least everyone involved :smile: )

Thank you all!

matthuhiggins commented:
It is possible to implement this in SQL, solving both the `order` & `offset` issues. A completely different approach that is easy to criticize due to the change in SQL generated, but at least it works without edge cases.

```ruby
def find_some(ids)
  result = order(connection.order_by_ids(ids)).where(primary_key => ids).to_a
  # ...
end

module ActiveRecord
  module ConnectionAdapters
    class MysqlAdapter
      def order_by_ids(ids)
        "FIELD(#{quote_column_name(primary_key)}, #{ids.join(',')})"
      end
    end
  end
end

module ActiveRecord
  module ConnectionAdapters
    class PostgreSQLAdapter
      def order_by_ids(ids)
        ids.map { |id| "ID=#{connection.quote(id)} DESC" }.join(',')
      end
    end
  end
end

fertapric commented:
> It is possible to implement this in SQL, solving both the order & offset issues. A completely different approach that is easy to criticize due to the change in SQL generated, but at least it works without edge cases.

@matthuhiggins I wouldn't go for the SQL implementation, it has a huge impact on performance; and this is a common query.

> I think we should simplify the approach for the case where find(array) and limit or offset are used at the same time. I'd simply pass the entire list of ids into the query and leave the database to determine the records to be returned.

As a regular user, I would not expect that behavior.

```ruby
Topic.find([3,2,5,1,4])
#=> [
  #<Topic id: 3, ...>,
  #<Topic id: 2, ...>,
  #<Topic id: 5, ...>,
  #<Topic id: 1, ...>,
  #<Topic id: 4, ...>
]

## Unexpected
Topic.offset(2).find([3,2,5,1,4])
#=> [
  #<Topic id: 3, ...>,
  #<Topic id: 2, ...>,
  #<Topic id: 4, ...>
]

## Expected
Topic.offset(2).find([3,2,5,1,4])
#=> [
  #<Topic id: 5, ...>,
  #<Topic id: 1, ...>,
  #<Topic id: 4, ...>
]
```

In addition, what would happen with `Topic.where(approved: true).limit(3).find([3,2,5,1,4])`?


_My two cents :grin:_
IMHO, for the sake of consistency, I would not include this functionality.
BUT, if you guys think ActiveRecord should be responsible of it, I would only include in-memory sorting with simple `#find`s and ignore it with `order`, `limit`, `offset` and so on.

matthuhiggins commented:
> I wouldn't go for the SQL implementation, it has a huge impact on performance

It has almost no impact on performance.

fertapric commented:
> > I wouldn't go for the SQL implementation, it has a huge impact on performance

> It has almost no impact on performance.

Maybe it's due to UUIDs, but these are the results I get:

```ruby
require "benchmark/ips"

n = 100
link_ids = Link.first(n).map(&:id)

Benchmark.ips do |x|
  x.report(:unordered) { Link.where(id: link_ids).to_a }
  x.report(:ordered) { Link.order(link_ids.map { |id| "ID='#{id}' DESC" }.join(',')).where(id: link_ids).to_a }
  x.compare!
end
```

### Queries

```ruby
>> Link.where(id: link_ids).to_a
  Link Load (0.8ms)  SELECT "links".* FROM "links" WHERE "links"."id" IN ('002178d6-a9ef-4b88-a20c-ddda0357fab6', '002bae32-539e-43ea-81b5-b3d5bf25639e')

>> Link.order(link_ids.map { |id| "ID='#{id}' DESC" }.join(',')).where(id: link_ids).to_a
  Link Load (1.2ms)  SELECT "links".* FROM "links" WHERE "links"."id" IN ('002178d6-a9ef-4b88-a20c-ddda0357fab6', '002bae32-539e-43ea-81b5-b3d5bf25639e')  ORDER BY ID='002178d6-a9ef-4b88-a20c-ddda0357fab6' DESC,ID='002bae32-539e-43ea-81b5-b3d5bf25639e' DESC
```

### n = 2

```
Calculating -------------------------------------
           unordered   120.000  i/100ms
             ordered   105.000  i/100ms
-------------------------------------------------
           unordered      1.204k (± 3.2%) i/s -      6.120k
             ordered      1.072k (± 2.9%) i/s -      5.460k

Comparison:
           unordered:     1204.4 i/s
             ordered:     1072.4 i/s - 1.12x slower
```

### n = 10

```
Calculating -------------------------------------
           unordered    81.000  i/100ms
             ordered    63.000  i/100ms
-------------------------------------------------
           unordered    809.350  (±12.0%) i/s -      3.888k
             ordered    647.350  (± 3.7%) i/s -      3.276k

Comparison:
           unordered:      809.3 i/s
             ordered:      647.4 i/s - 1.25x slower
```

### n = 20

```
Calculating -------------------------------------
           unordered    63.000  i/100ms
             ordered    40.000  i/100ms
-------------------------------------------------
           unordered    631.418  (± 4.3%) i/s -      3.213k
             ordered    399.571  (± 3.5%) i/s -      2.000k

Comparison:
           unordered:      631.4 i/s
             ordered:      399.6 i/s - 1.58x slower
```

### n = 100

```text
Calculating -------------------------------------
           unordered    21.000  i/100ms
             ordered     2.000  i/100ms
-------------------------------------------------
           unordered    209.608  (± 5.7%) i/s -      1.050k
             ordered     23.893  (± 0.0%) i/s -    120.000

Comparison:
           unordered:      209.6 i/s
             ordered:       23.9 i/s - 8.77x slower
```

miguelgraz committed:
Adding a new test using chained where, limit and find([pks])

miguelgraz commented:
Hello again folks!

I'd like to point that the current state of this PR (seems to) handle `limit`, `order` and `where` flawlessly, the only problem/doubt is with `offset`.

So, I'm not an expert but _my two cents_:

@matthuhiggins I wouldn't go with raw SQL on this case, from my point of view it'd change too much the approach and perhaps it'd add the need to write a custom method for each supported DB, please correct me if I'm wrong here.

@fertapric I just added [another test case](https://github.com/rails/rails/pull/20453/files#diff-de79d7a2a56f1ed515efcb2381baa08aR103) to use the sample code `Topic.where(approved: true).limit(3).find([3,2,5,1,4])` you asked about, it works fine but you'll notice that I've added a comment, turns out that if you filter and then try to `find` a filtered or nonexistent record it raises a `ActiveRecord::RecordNotFound` error, seems correct to me, what do you think?

Again from my point of view we have two (maybe three) possible approaches on this problem with `offset`:

1. Just let the database select which records should be returned, as @jmbejar suggested (and here we could add the support to `limit` or not);
2. Just remove the feature altogether as @fertapric suggested;
3. Find a smarter way to handle every case it already does _and_ `offset`?

What do you guys think?

Thanks again everyone! :snowboarder:

matthuhiggins commented:
> I wouldn't go with raw SQL on this case, from my point of view it'd change too much the approach and perhaps it'd add the need to write a custom method for each supported DB

You are correct in that it's a big change in the approach. The implementation is fewer lines of code than you've written.

@fertapric's performance analysis is based on a database without data. His test is testing string creation and query parsing performance, rather than the cost of ordering data in the database vs in ruby.

miguelgraz commented:
Would it require a different implementation for each database? Don't get me
wrong, if we find that this is a cheap trade-off I'm all for it, guess we
have another option to consider then, and that's great :)

On Fri, Aug 7, 2015 at 9:33 PM, Matthew Higgins <notifications@github.com>
wrote:

> I wouldn't go with raw SQL on this case, from my point of view it'd change
> too much the approach and perhaps it'd add the need to write a custom
> method for each supported DB
>
> You are correct in that it's a big change in the approach. The
> implementation is fewer lines of code than you've written.
>
> @fertapric <https://github.com/fertapric>'s performance analysis is based
> on a database without data. His test is testing string creation and query
> parsing performance, rather than the cost of ordering data in the database
> vs in ruby.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rails/rails/pull/20453#issuecomment-128869376>.
>



-- 
Miguel Grazziotin
Miguelgraz.com

miguelgraz commented:
@matthuhiggins I've edited my previous comment with our possible options to add your approach :)

jmbejar commented:
Probably the best option is what @fertapric suggested: removing the feature (in other words, ignoring #20338). I doubt if this is something most developers would expect from the `find` method.

In the other hand, if we still want to sort the results based on the array of values passed to the `find` method, I would let the database behave like it does with `limit` and `offset`, just making sure that results are ordered based on the input array.

miguelgraz commented:
I think it'd be a nice-to-have feature instead of just retrieving it using IDs order like `find` works today, even if we have to let the database handle the `offset` clause it seems like an improvement to me.

rafaelfranca commented:
r? @rafaelfranca

miguelgraz commented:
:heart:

jmbejar commented:
@miguelgraz :clap:

miguelgraz commented:
@jmbejar unfortunately I couldn't achieve every desired case and @matthewd had to spend some time helping with these edge cases but still I'm very glad to help at least a little bit and this was only the first PR, hopefully next ones will be better :smile: 

:snowboarder:

Najaf commented:
:+1:

nettofarah commented:
yay, @miguelgraz
